src/indexing.c

```diff
@@ -301,3 +304,67 @@ ts_indexing_find_clustered_index(Oid table_relid)
 
 	return index_relid;
 }
+
+typedef enum IndexValidity
+{
+	IndexInvalid = 0,
+	IndexValid,
+} IndexValidity;
+
+static bool
+ts_indexing_mark_as(Oid index_id, IndexValidity validity)
+{
+	Relation pg_index;
+	HeapTuple indexTuple;
+	HeapTuple new_tuple;
+	Form_pg_index indexForm;
+	bool was_valid;
+
+	/* Open pg_index and fetch a writable copy of the index's tuple */
+	pg_index = heap_open(IndexRelationId, RowExclusiveLock);
+
+	indexTuple = SearchSysCacheCopy1(INDEXRELID, ObjectIdGetDatum(index_id));
+
+	if (!HeapTupleIsValid(indexTuple))
+		elog(ERROR, "cache lookup failed when marking index %u", index_id);
+
+	new_tuple = heap_copytuple(indexTuple);
+	indexForm = (Form_pg_index) GETSTRUCT(new_tuple);
+
+	was_valid = indexForm->indisvalid;
+
+	/* Perform the requested state change on the copy */
+	switch (validity)
+	{
+		case IndexValid:
+			Assert(indexForm->indislive);
+			Assert(indexForm->indisready);
+			indexForm->indisvalid = true;
+			break;
+		case IndexInvalid:
+			indexForm->indisvalid = false;
+			indexForm->indisclustered = false;
+			break;
+		default:
+			elog(ERROR, "internal error");
+	}
+
+	/* ... and write it back (unlike the original, not in-place) */
```

Comment:
writing in-place comes with a whole bunch of constraints about when you're allowed to do it, and, since we have no need to perform an in-place update, we might as well perform a normal update and have the function be less brittle.
The not in-place _is_ what's different, though I think I'll going to remove the second half of that comment, this function is different enough from the original at this point that it is not worth the mention.